software/image_processor.py

```diff
@@ -0,0 +1,180 @@
+                color_frame = [],
+                depth_frame = [],
+                fragmented = [],
+                debug_frame = []) -> None:
+
+
+        self.balls = balls
+        self.basket_b = basket_b
+        self.basket_m = basket_m
+        self.color_frame = color_frame
+        self.depth_frame = depth_frame
+        self.fragmented = fragmented
+
+        # can be used to illustrate things in a separate frame buffer
+        self.debug_frame = debug_frame
+
+
+#Main processor class. processes segmented information
+class ImageProcessor():
+    def __init__(self, camera, color_config = "colors.pkl", debug = False):
+        self.camera = camera
+        
+        self.color_config = color_config
+        with open(self.color_config, 'rb') as conf:
+            self.colors_lookup = pickle.load(conf)
+            self.set_segmentation_table(self.colors_lookup)
+
+        self.fragmented	= np.zeros((self.camera.rgb_height, self.camera.rgb_width), dtype=np.uint8)
+
+        self.t_balls = np.zeros((self.camera.rgb_height, self.camera.rgb_width), dtype=np.uint8)
+        self.t_basket_b = np.zeros((self.camera.rgb_height, self.camera.rgb_width), dtype=np.uint8)
+        self.t_basket_m = np.zeros((self.camera.rgb_height, self.camera.rgb_width), dtype=np.uint8)
+
+        self.debug = debug
+        self.debug_frame = np.zeros((self.camera.rgb_height, self.camera.rgb_width), dtype=np.uint8)
+
+    def set_segmentation_table(self, table):
+        segment.set_table(table)
+
+    def start(self):
+        self.camera.open()
+
+    def stop(self):
+        self.camera.close()
+
+    def analyze_balls(self, t_balls, fragments) -> list:
+        #dilation here
+        contours, hierarchy = cv2.findContours(t_balls, cv2.RETR_EXTERNAL, cv2.CHAIN_APPROX_SIMPLE)
+
+        balls = []
+
+        for contour in contours:
+
+            # ball filtering logic goes here. Example includes filtering by size and an example how to get pixels from
+            # the bottom center of the fram to the ball
+
+            size = cv2.contourArea(contour)
+
+            if size < 15:
+                continue
+
+            x, y, w, h = cv2.boundingRect(contour)
+
+            ys	= np.array(np.arange(y + h, self.camera.rgb_height), dtype=np.uint16)
+            xs	= np.array(np.linspace(x + w/2, self.camera.rgb_width / 2, num=len(ys)), dtype=np.uint16)
+
+            #testobjdist, testobjx, testobjy = position(h,w,depth,x,y)
+            obj_x = int(x + (w/2))
+            obj_y = int(y + (h/2))
+            obj_dst = obj_y
+
+            if self.debug:
+                self.debug_frame[ys, xs] = [0, 0, 0]
+                cv2.circle(self.debug_frame,(obj_x, obj_y), 10, (0,255,0), 2)
+
+            balls.append(Object(x = obj_x, y = obj_y, size = size, distance = obj_dst, exists = True))
+        #[Object: x=7; y=212; size=15.0; distance=212; exists=True]
+        
+        #balls.sort(key= lambda x: x.distance)

```

Comment:
There is now no sorting. Is this intentional?

Reply:
Wrote it back in.